src/dbnode/storage/bootstrap/bootstrapper/readers.go

```diff
@@ -0,0 +1,242 @@
+		if err := result.Err.Error(); err != nil {
+			logger.Error("fs bootstrapper unable to read info file",
+				zap.Uint32("shard", shard),
+				zap.Stringer("namespace", ns.ID()),
+				zap.Error(err),
+				zap.String("timeRange", tr.String()),
+				zap.String("path", result.Err.Filepath()),
+			)
+			// Errors are marked unfulfilled by markRunResultErrorsAndUnfulfilled
+			// and will be re-attempted by the next bootstrapper.
+			continue
+		}
+
+		info := result.Info
+		blockStart := xtime.FromNanoseconds(info.BlockStart)
+		if !tr.Overlaps(xtime.Range{
+			Start: blockStart,
+			End:   blockStart.Add(ns.Options().RetentionOptions().BlockSize()),
+		}) {
+			// Errors are marked unfulfilled by markRunResultErrorsAndUnfulfilled
+			// and will be re-attempted by the next bootstrapper.
+			continue
+		}
+
+		r, err := readerPool.Get()
+		if err != nil {
+			logger.Error("unable to get reader from pool")
+			// Errors are marked unfulfilled by markRunResultErrorsAndUnfulfilled
+			// and will be re-attempted by the next bootstrapper.
+			continue
+		}
+
+		openOpts := fs.DataReaderOpenOptions{
+			Identifier: fs.FileSetFileIdentifier{
+				Namespace:  ns.ID(),
+				Shard:      shard,
+				BlockStart: blockStart,
+			},
+		}
+		if err := r.Open(openOpts); err != nil {
+			logger.Error("unable to open fileset files",
+				zap.Uint32("shard", shard),
+				zap.Time("blockStart", blockStart),
+				zap.Error(err),
+			)
+			readerPool.Put(r)
+			// Errors are marked unfulfilled by markRunResultErrorsAndUnfulfilled
+			// and will be re-attempted by the next bootstrapper.
+			continue
+		}
+
+		readers = append(readers, r)
+	}
+
+	return ShardReaders{Readers: readers}
+}
+
+// ReaderPool is a lean pool that does not allocate
+// instances up front and is used per bootstrap call.
+type ReaderPool struct {
+	sync.Mutex
+	alloc        ReaderPoolAllocFn
+	values       []fs.DataFileSetReader
+	disableReuse bool
+}
+
+// ReaderPoolAllocFn allocates a new fileset reader.
+type ReaderPoolAllocFn func() (fs.DataFileSetReader, error)
+
+// NewReaderPoolOptions contains reader pool options.
+type NewReaderPoolOptions struct {
+	Alloc        ReaderPoolAllocFn
+	DisableReuse bool
+}
+
+// NewReaderPool creates a new share-able fileset reader pool
+func NewReaderPool(
+	opts NewReaderPoolOptions,
+) *ReaderPool {
+	return &ReaderPool{alloc: opts.Alloc, disableReuse: opts.DisableReuse}
+}
+
+// Get gets a fileset reader from the pool in synchronized fashion.
+func (p *ReaderPool) Get() (fs.DataFileSetReader, error) {
+	p.Lock()
+	if len(p.values) == 0 {
+		p.Unlock()
+		return p.alloc()
+	}
+	length := len(p.values)
+	value := p.values[length-1]
+	p.values[length-1] = nil
+	p.values = p.values[:length-1]
+	p.Unlock()
+	return value, nil
+}
+
+// Put returns a fileset reader back the the pool in synchronized fashion.
+func (p *ReaderPool) Put(r fs.DataFileSetReader) {
+	if p.disableReuse {
+		// Useful for tests.
+		return
+	}
+	p.Lock()
```

Comment:
Somewhat related commentary: Hopefully with Go 1.14 we can just use `defer` everywhere to unlock locks with pretty much zero cost: https://twitter.com/bradfitz/status/1184547163235184640?lang=en